File: 2022/day04/zig/solution.zig
const std = @import("std");

const Range = struct {
    start: u32,
    end: u32,
};

const Pair = struct {
    first: Range,
    second: Range,
};

fn fullyContains(pair: Pair) bool {
    // Check if one range fully contains the other
    const first_contains_second = pair.first.start <= pair.second.start and pair.first.end >= pair.second.end;
    const second_contains_first = pair.second.start <= pair.first.start and pair.second.end >= pair.first.end;
    return first_contains_second or second_contains_first;
}

fn overlaps(pair: Pair) bool {
    // Check if ranges overlap at all
    return pair.first.start <= pair.second.end and pair.second.start <= pair.first.end;
}

fn parseRange(s: []const u8) !Range {
    var iter = std.mem.splitScalar(u8, s, '-');
    const start_str = iter.next() orelse return error.InvalidFormat;
    const end_str = iter.next() orelse return error.InvalidFormat;
    return Range{
        .start = try std.fmt.parseInt(u32, start_str, 10),
        .end = try std.fmt.parseInt(u32, end_str, 10),
    };
}

fn parseLine(line: []const u8) !Pair {
    var iter = std.mem.splitScalar(u8, line, ',');
    const first_str = iter.next() orelse return error.InvalidFormat;
    const second_str = iter.next() orelse return error.InvalidFormat;
    return Pair{
        .first = try parseRange(first_str),
        .second = try parseRange(second_str),
    };
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Parse pairs and count results
    var part1_count: u32 = 0;
    var part2_count: u32 = 0;

    var lines = std.mem.splitScalar(u8, content, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        const pair = try parseLine(line);

        if (fullyContains(pair)) {
            part1_count += 1;
        }
        if (overlaps(pair)) {
            part2_count += 1;
        }
    }

    const stdout = std.fs.File.stdout();
    var buffer: [256]u8 = undefined;
    var file_writer = stdout.writer(&buffer);
    const writer = &file_writer.interface;
    try writer.print("Part 1: {d}\n", .{part1_count});
    try writer.print("Part 2: {d}\n", .{part2_count});
    try writer.flush();
}

File: 2022/day04/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;

# Get script directory and input file path
use File::Basename;
use File::Spec;

my $script_dir = dirname(__FILE__);
my $input_file = File::Spec->catfile($script_dir, '..', 'input.txt');

# Parse input and compute results
my ($part1, $part2) = (0, 0);

open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
while (my $line = <$fh>) {
    chomp $line;
    next unless $line;

    # Parse line like "2-4,6-8"
    my ($a1, $b1, $a2, $b2) = $line =~ /(\d+)-(\d+),(\d+)-(\d+)/;

    # Part 1: Check if one range fully contains the other
    if (($a1 <= $a2 && $b1 >= $b2) || ($a2 <= $a1 && $b2 >= $b1)) {
        $part1++;
    }

    # Part 2: Check if ranges overlap at all
    if ($a1 <= $b2 && $a2 <= $b1) {
        $part2++;
    }
}
close $fh;

print "Part 1: $part1\n";
print "Part 2: $part2\n";
